### PR TITLE
feat(new transform): Add swimlanes transform

### PR DESCRIPTION
**Pull request by Jeffail**
This adds a `swimlanes` transform, allowing you to easily organize forks in your topology. Each `lane` is a named `condition`, which is an internal component type not yet exposed to users, with one single implementation `check_fields`. Conditions have already had a spec written within the unit testing work: https://vector.dev/docs/reference/tests/#conditions.

The implementation here works by allowing a `swimlanes` transform config to divide itself into N configurations before building our topology. Each lane becomes its own transform config, where it essentially acts as a basic filter with a name of the form `<swimlanes_name>.<lane_name>`, it's inputs are a copy of the inputs of the parent `swimlanes`.

This implementation is fairly straight forward as we don't need to accommodate fanning out our transforms at the topology execution level. However, this implementation is also limited due to the expanded children of a transform being independent, each with their own feed of messages. This means we _could_ support an `if` or `switch` transform, but it's less efficient than consuming one event and outputting across N channels.

Closes https://github.com/timberio/vector/issues/1724

### CODE REVIEW

**Comment by binarylogic** [comment]
👍 I'm excited for this one.

**Comment by binarylogic** [comment]
@Jeffail tag @LucioFranco and @Hoverbear when you're ready for a review.

File: src/topology/unit_test.rs [hunk 1]
```diff
@@ -597,13 +611,73 @@ mod tests {
...
-        let mut tests = build_unit_tests(&config).unwrap();
+        let mut tests = build_unit_tests(&mut config).unwrap();
+        assert_eq!(tests[0].run().1, Vec::<String>::new());
+    }
+
+    #[test]
+    fn test_swimlanes() {
+        let mut config: Config = toml::from_str(
+            r#"
+[transforms.foo]
```
Comment (binarylogic) [inline src/topology/unit_test.rs]: I'm curious if you think this is better served as a behavior test? https://github.com/timberio/vector/pull/1781

File: src/topology/config/mod.rs [hunk 2]
```diff
@@ -194,6 +194,13 @@ pub struct TransformOuter {
 
 #[typetag::serde(tag = "type")]
 pub trait TransformConfig: core::fmt::Debug {
+    /// Allows a transform configuration to expand itself into multiple "child"
+    /// transformations to replace it. This allows a transform to act as a macro
+    /// for various patterns.
+    fn expand(&mut self) -> Option<IndexMap<String, Box<dyn TransformConfig>>> {
```
Comment (lukesteensen) [inline src/topology/config/mod.rs]: This is really cool! I've been thinking about a way to make an expansion/optimization pass on configs before actually building them and this seems like a nice and simple way to start experimenting with that.

File: src/topology/config/mod.rs [hunk 3]
```diff
@@ -299,6 +306,49 @@ impl Config {
...
+        let mut expanded_transforms = IndexMap::new();
+        let mut expansions = IndexMap::new();
+        let mut errors = Vec::new();
+
+        while let Some((k, mut t)) = self.transforms.pop() {
+            if let Some(expanded) = t.inner.expand() {
+                let mut children = Vec::new();
+                for (name, mut child) in expanded {
+                    let full_name = format!("{}.{}", k, name);
+                    if !child.expand().is_none() {
```
Comment (lukesteensen) [inline src/topology/config/mod.rs]: Nitpick: could just use `is_some` here.

File: src/transforms/swimlanes.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                    name: k,
+                    condition: v,
+                }),
+            );
+        }
+
+        if !map.is_empty() {
+            Some(map)
+        } else {
+            None
```
Comment (lukesteensen) [inline src/transforms/swimlanes.rs]: Would this case lead to the error in `build` above? Not sure the best place to raise this to the user.

**Comment by binarylogic** [comment]
Also, could we add [behavior tests](https://github.com/timberio/vector/tree/master/tests/behavior/transforms) for this? Specifically around the use of nested fields so we don't break anything in https://github.com/timberio/vector/pull/1662.

File: src/transforms/swimlanes.rs (hunk 4 above)
Comment (Jeffail) [reply src/transforms/swimlanes.rs]: Was a little torn on this one, in the end I've tried out `crate::Result<Option<_>>`, seems to be the worst of evils. There will definitely be cases where expanding transforms might fail and I agree it's dirty to return `None` and then expose the error in `build`.

**Comment by Jeffail** [comment]
Ready for review now @binarylogic @LucioFranco @Hoverbear.

**Comment by binarylogic** [comment]
@Jeffail I added examples to the docs and split out the "conditions" to `.meta/_partials/_conditions.toml`. Can you review that and make any changes necessary?

**Comment by binarylogic** [comment]
I just noticed your "Creating Swimlanes" section. I'm wondering if we should just have a common "Examples" or "Demos" section for this kind of stuff. We have it for many other components as well.

**Comment by Hoverbear** [comment]
Whoops...

**Comment by Hoverbear** [comment]
This generally looks good to me but I'd like to see it conflict free and passing all checks before a merge. :)

File: src/transforms/swimlanes.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "swimlanes"
+    }
+}
+
+//------------------------------------------------------------------------------
```
Comment (LucioFranco) [inline src/transforms/swimlanes.rs]: Is there a reason there are no tests for this transform? I see there are some in unit_test but may be worth while to at least add a happy path test here?

File: src/transforms/swimlanes.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+//------------------------------------------------------------------------------
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct SwimlaneConfig {
+    #[serde(skip)]
+    name: String,
```
Comment (LucioFranco) [inline src/transforms/swimlanes.rs]: I don't quite see why this is included? I don't see it being used?

File: src/transforms/swimlanes.rs (hunk 5 above)
Comment (Hoverbear) [reply src/transforms/swimlanes.rs]: There are some behavior tests, but those sadly aren't part of a `cargo test` workflow.

**Pull request merged**